### DISCUSSION TITLE
OpenVINO and ONNX support for faster CPU execution

### DISCUSSION THREAD

**Discussion by zhuzilin** [openai/whisper#208] [Show and tell]
Hi! Thank you so much for this amazing project!

I've been playing with the model and find them work terrific, even for the "tiny" model, so I transformed the models into openvino and onnx, and organized them in [github.com/zhuzilin/whisper-openvino](https://github.com/zhuzilin/whisper-openvino) and [huggingface hub](https://huggingface.co/models?search=whisper-openvino), so that people could give them a faster try without GPU and make it easier to deploy on client side.

During my project of exporting models into openvino, I got 40% end-to-end time reduction (a 90-minute audio takes 40 minutes now instead of 68 minutes). But I believe there is still large potential in improving the CPU performance of whisper models. Therefore, I'm posting it here so that people can give it a try and help me to make it better :)

**Comment by ggerganov** [comment]
Hi, nice work!

I was immediately interested in your result, because I am working on my own custom inference implementation of Whisper for CPU and I am focusing on efficiency and low-memory usage. I just ran the same transcription that you did with the [1h 30min Cramack video](https://www.youtube.com/watch?v=P6UKhR0T6cs) and I got the following results:

| Model | Time to transcribe | Result |
| ---      | ---                           | ---       |
| tiny.en | 116 sec | [carmack.wav-tiny.en.txt](https://github.com/openai/whisper/files/9686673/carmack.wav-tiny.en.txt) |
| small.en | 563 sec | [carmack.wav-small.en.txt](https://github.com/openai/whisper/files/9686680/carmack.wav-small.en.txt) |

This is running on a MacBook M1 Pro - CPU only using 8 threads.

My implementation is available here:

https://github.com/ggerganov/whisper.cpp

If you are interested, you can give it a try and see how it performs using your hardware.

---

Edit:
I tried installing `whisper-openvino` on my MacBook by get the following error:
```
$ pip install git+https://github.com/zhuzilin/whisper-openvino.git
Collecting git+https://github.com/zhuzilin/whisper-openvino.git
  Cloning https://github.com/zhuzilin/whisper-openvino.git to /private/var/folders/fr/8wtnvzqx0h73rmq93vmtbv380000gn/T/pip-req-build-r430m0wd
  Running command git clone -q https://github.com/zhuzilin/whisper-openvino.git /private/var/folders/fr/8wtnvzqx0h73rmq93vmtbv380000gn/T/pip-req-build-r430m0wd
  Resolved https://github.com/zhuzilin/whisper-openvino.git to commit 7c38500b72d8b6ed245e59be4b303a66c56f93e0
Requirement already satisfied: numpy in ./lib/python3.10/site-packages (from whisper==1.0) (1.23.3)
Requirement already satisfied: torch in ./lib/python3.10/site-packages (from whisper==1.0) (1.12.1)
Requirement already satisfied: tqdm in ./lib/python3.10/site-packages (from whisper==1.0) (4.64.1)
Requirement already satisfied: more-itertools in ./lib/python3.10/site-packages (from whisper==1.0) (8.14.0)
Requirement already satisfied: transformers>=4.19.0 in ./lib/python3.10/site-packages (from whisper==1.0) (4.22.2)
Requirement already satisfied: ffmpeg-python==0.2.0 in ./lib/python3.10/site-packages (from whisper==1.0) (0.2.0)
ERROR: Could not find a version that satisfies the requirement openvino (from whisper) (from versions: none)
ERROR: No matching distribution found for openvino
```
Maybe on Arm there are some extra steps?

**Reply by ArtyomZemlyak** [reply]
Maybe its can help:
https://docs.openvino.ai/latest/openvino_docs_install_guides_installing_openvino_from_archive_macos.html#doxid-openvino-docs-install-guides-installing-openvino-from-archive-macos

**Reply by zhuzilin** [reply]
Could you try `pip install openvino` on your machine? Because I'm not sure if openvino is supported on M1 (it is not on python 3.10 or higher, btw).

And the performance you got is amazing! I just rerun my code with your config (using .wav of sample rate 16kHz and greedy decoder instead of beam search) and it still takes 16 minutes. However, I failed to compile and run your code, the output is:
```bash
zilinzhu@MacBook-Pro-3 ~/D/whisper.cpp (master) [2]> ./main -m models/ggml-tiny.en.bin -f ~/Documents/whisper-openvino/carmack.wav        (base) 
whisper_model_load: loading model from 'models/ggml-tiny.en.bin'
whisper_model_load: n_vocab       = 51864
whisper_model_load: n_audio_ctx   = 1500
whisper_model_load: n_audio_state = 384
whisper_model_load: n_audio_head  = 6
whisper_model_load: n_audio_layer = 4
whisper_model_load: n_text_ctx    = 448
whisper_model_load: n_text_state  = 384
whisper_model_load: n_text_head   = 6
whisper_model_load: n_text_layer  = 4
whisper_model_load: n_mels        = 80
whisper_model_load: f16           = 1
whisper_model_load: type          = 1
whisper_model_load: mem_required  = 244.00 MB
whisper_model_load: adding 1607 extra tokens
whisper_model_load: ggml ctx size =  84.99 MB
whisper_model_load: memory size =    11.41 MB 
whisper_model_load: model size  =    73.54 MB
fish: Job 1, './main -m models/ggml-tiny.en.b…' terminated by signal SIGSEGV (Address boundary error)
```
Do you know what could be the reason of it? The wav file is generated from the origin mp3 file with `ffmpeg -i carmack.mp3 -ar 16000 carmack.wav `

**Reply by zhuzilin** [reply]
And also, in the current `whisper-openvino`, the models are all fp32, that's maybe part of the reason why it's not that fast.

**Reply by ArtyomZemlyak** [reply]
Can you try this for fp16? (i dont tried, but mb it works) 
https://docs.openvino.ai/latest/openvino_docs_MO_DG_FP16_Compression.html

**Reply by ggerganov** [reply]
@zhuzilin There indeed was an issue when using stereo WAV files. You can either convert it to mono audio using `ffmpeg -i carmack.mp4 -ar 16000 -ac 1 -c:a pcm_s16le carmack.wav` or pull the latest `whisper.cpp` - it should be fixed.

And yes - you have correctly figured out to use the greedy sampling strategy to make the comparison fair.

Here is the output of `pip install openvino`:

```
$ pip install openvino
ERROR: Could not find a version that satisfies the requirement openvino (from versions: none)
ERROR: No matching distribution found for openvino
```
I'm using Python 3.10.2

Will check the link from @ArtyomZemlyak in the following days and see if I can make it work.

**Reply by richardburleigh** [reply]
@ggerganov use `pip install openvino-arm`. Pip will get an error due to the name mismatch, so also remove openvino from the requirements

**Reply by ggerganov** [reply]
@richardburleigh `pip install openvino-arm` worked but I get the following error when I try to run the example:

```python
$ whisper --beam_size 0 --best_of 1 --model tiny.en tests/jfk.flac 

lib/python3.10/site-packages/whisper/transcribe.py:272: UserWarning: tiny.en is an English-only model but receipted 'None'; using English instead.
  warnings.warn(f"{model_name} is an English-only model but receipted '{args['language']}'; using English instead.")
Traceback (most recent call last):
  File "bin/whisper", line 33, in <module>
    sys.exit(load_entry_point('whisper==1.0', 'console_scripts', 'whisper')())
  File "lib/python3.10/site-packages/whisper/transcribe.py", line 283, in cli
    model = load_model(model_name)
  File "lib/python3.10/site-packages/whisper/__init__.py", line 104, in load_model
    model = Whisper(dims, name)
  File "lib/python3.10/site-packages/whisper/model.py", line 77, in __init__
    self.decoder = OpenVinoTextDecoder(model=model)
  File "lib/python3.10/site-packages/whisper/model.py", line 55, in __init__
    self.model = self.core.compile_model(self._model, "CPU")
  File "lib/python3.10/site-packages/openvino/runtime/ie_api.py", line 266, in compile_model
    super().compile_model(model, device_name, {} if config is None else config)
RuntimeError: get_shape was called on a descriptor::Tensor with dynamic shape
```

Any guess what I am doing wrong?

**Reply by richardburleigh** [reply]
My apologies @ggerganov - It turns out that is an old version of openvino that is not compatible.
Based on the [response by @ArtyomZemlyak ](https://github.com/openai/whisper/discussions/208#discussioncomment-3780625), I needed to build openvino from source as per [these instructions](https://github.com/openvinotoolkit/openvino/wiki/BuildingForMacOS_arm64).

**Reply by ArtyomZemlyak** [reply]
Dynamic shapes not supported in old versions OpenVino

**Reply by zhuzilin** [reply]
> There indeed was an issue when using stereo WAV files. You can either convert it to mono audio using ffmpeg -i carmack.mp4 -ar 16000 -ac 1 -c:a pcm_s16le carmack.wav or pull the latest whisper.cpp - it should be fixed.

Sorry I was on vacation the last week. On my laptop, it only takes less than 6 minutes to transcribe the whole audio (versus 16 mins on my version). That's amazing!

**Reply by genevera** [reply]
@ggerganov This is _awesome_ and it runs orders of magnitude faster than the stock version of whisper on my M1

**Reply by jafri** [reply]
@ggerganov running on M1 Max machine, can do it in 406 seconds for small.en with 10 threads

If I split the video into 5 chunks of 17 minutes each, and run 5 concurrent instances with 2 threads each, can run it in 220 seconds

If I split the video into 10 chunks of 8.5 minutes each, and run 10 concurrent instances with 1 threads each, can run it in 180 seconds

**Comment by ArtyomZemlyak** [comment]
Tried converting onnx models to fp16 OpenVino models. Like this:

```sh
mo --input_model ../models/openvino/medium/decoder.onnx --data_type FP16
```
But cant see any improvements on my task, compared to your OpenVino models (10s / 200 s audio test):

| Model| T, s | CPU/GPU |
| :------------- | --------------- | :-------- |
| tiny         | 10            | CPU     |
| tiny_fp16    | 10            | CPU     |
| base         | 2 / 122       | CPU     |
| base_fp16    | 2 / 130       | CPU     |
| medium       | 41            | CPU     |
| medium_fp16  | 41            | CPU     |

**Comment by ArtyomZemlyak** [comment]
And can you check your model files on HuggingFace. I tried small model, but has error about shapes - seems its base model. Maybe in HF not appropriate models loaded.

**Reply by rjwilmsi** [reply]
Yes, I have found same problem. Tiny.en and base.en models fine, small.en errors on same input file:
...
  File "/usr/local/lib/python3.8/site-packages/openvino-2022.2.0-py3.8-linux-x86_64.egg/openvino/runtime/ie_api.py", line 267, in infer_new_request
    return self.create_infer_request().infer(inputs)
  File "/usr/local/lib/python3.8/site-packages/openvino-2022.2.0-py3.8-linux-x86_64.egg/openvino/runtime/ie_api.py", line 147, in infer
    return super().infer(normalize_inputs(self, inputs))
RuntimeError: [ PARAMETER_MISMATCH ] Can not clone with new dims. Descriptor's shape: {48, 0 - ?, 0 - ?, 1024} is incompatible with provided dimensions: {24, 5, 1, 768}.

@zhuzilin Does the small.en model have an issue?

**Reply by whalefa1I** [reply]
RuntimeError: [ PARAMETER_MISMATCH ] Can not clone with new dims. Descriptor's shape: {12, 0 - ?, 0 - ?, 512} is incompatible with provided dimensions: {24, 3, 3, 768}.
Almost same error for "small" model @zhuzilin

**Reply by carlosgalveias** [reply]
Same error using the small model:
RuntimeError: [ PARAMETER_MISMATCH ] Can not clone with new dims. Descriptor's shape: {12, 0 - ?, 0 - ?, 512} is incompatible with provided dimensions: {24, 5, 3, 768}.
Any tips on how to solve this?

**Comment by ArtyomZemlyak** [comment]
Additional testing for tiny model. I think, 

- Cant create dummy_input without kv_cache for onnx and openvino.
- Speedup of torch decoder with kv cache more then 5 times.

### Encoder
TIME torch: 2.716999053955078

TIME onnxruntime: 2.300724983215332     zhuzilin
TIME onnxruntime: 2.098496198654175     zhuzilin QUANT

TIME openvino: 1.8365371227264404   zhuzilin
TIME openvino: 1.8365371227264404   zhuzilin bin_xml
TIME openvino: 1.8805508613586426   zhuzilin fp16

TIME deepsparse: 1.354171991348266  zhuzilin
TIME deepsparse: 1.354171991348266  zhuzilin    QUANT

### Decoder

TIME torch: 1.5086073875427246
TIME torch: 0.2621121406555176      kv_cache
TIME openvino: 0.28642868995666504  zhuzilin
TIME openvino: 0.2409522533416748   zhuzilin    fp16
TIME onnxruntime: 0.21176910400390625   zhuzilin

**Reply by Majdoddin** [reply]
Is it [deepsparse ](https://neuralmagic.com/deepsparse-engine/) from neuralmagic?

**Reply by ArtyomZemlyak** [reply]
Yes, it is! @Majdoddin

**Reply by Majdoddin** [reply]
Great 👍 
Do you know of any attempt to apply [SparseML ](https://neuralmagic.com/sparseml/) to whisper to sparsify its model? I am considering doing it.

**Reply by ArtyomZemlyak** [reply]
Cant realy help with this question, bc i dont deep dive in this framework.
But I think, you need avoid methods, which make influence on model weights. Especially for non-en languages. Its because whisper has dependency of his inference speed from how good he can transcribe file in raw default mode without any greedy decoder or beem search.

**Comment by ArtyomZemlyak** [comment]
I think that it is the place to be in parallel to test the speed of the encoder and decoder separately. Since it can show a net gain in performance.

If we take measurements of the entire process of audio recognition, then everything depends very much on the quality of the audio, how legible the speech is, of the selected TokenDecoder.

If you look at the results here #212 , you can see that in some cases, heavier models are processed faster - this is because they almost immediately give a normal result. While a smaller model needs more passes to get a more or less normal result.

**Reply by tom-huntington** [reply]
> If you look at the results here #212 , you can see that in some cases, heavier models are processed faster - this is because they almost immediately give a normal result. While a smaller model needs more passes to get a more or less normal result.

I'm confused. Even the GreedyDecoder shows a speed up for larger models, so where are the extra passes coming from?

**Reply by tom-huntington** [reply]
> so where are the extra passes coming from?

https://github.com/openai/whisper/blob/02aa851a4910201f0db56960064d7e121a01002c/whisper/transcribe.py#L102-L128

**Comment by richardburleigh** [comment]
Amazing work, but there seems to be a problem with the models on Huggingface:
```
>>> whisper.load_model("base")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/lib/python3.10/site-packages/whisper/__init__.py", line 104, in load_model
    model = Whisper(dims, name)
  File "/opt/homebrew/lib/python3.10/site-packages/whisper/model.py", line 77, in __init__
    self.decoder = OpenVinoTextDecoder(model=model)
  File "/opt/homebrew/lib/python3.10/site-packages/whisper/model.py", line 55, in __init__
    self.model = self.core.compile_model(self._model, "CPU")
  File "/opt/homebrew/lib/python3.10/site-packages/openvino/runtime/ie_api.py", line 266, in compile_model
    super().compile_model(model, device_name, {} if config is None else config)
RuntimeError: get_shape was called on a descriptor::Tensor with dynamic shape
```

**Reply by ArtyomZemlyak** [reply]
What version of openvino are you using? Try lastest (>=2022.1).

**Reply by richardburleigh** [reply]
Thank you @ArtyomZemlyak! I needed to build from source and it worked. Massive performance improvements using your code.

**Comment by MaxwellDPS** [comment]
Great work @zhuzilin! Tagging this here as it seems relevant https://github.com/openai/whisper/discussions/227 . Same issue with your code base as in #227

**Comment by zara0m** [comment]
Hello,
It was amazing!

I am trying to create Onnx files myself, could you please guide me how to do that? For example how did you write the Onnx Configuration?

Thank you

**Reply by Kirankumar2609** [reply]
Hi @zara0m. I am trying to do the same. Care to share some findings?

**Comment by Y-T-G** [comment]
@zhuzilin Did you save the ONNX models that you used to convert to OpenVINO anywhere?

**Reply by zhuzilin** [reply]
They are all uploaded to huggingface.

**Reply by Y-T-G** [reply]
@zhuzilin Thanks. I found them.

**Reply by Y-T-G** [reply]
@zhuzilin Was [this](https://github.com/zhuzilin/whisper-openvino/blob/edb69443a541955b849a3743d567910eb023597a/whisper/decoding.py#L152) the code you used for the decoder export?

**Reply by abhinavkulkarni** [reply]
@zhuzilin: Can you please tell us how to use the models you uploaded on HF using command line `whisper`? For e.g., when I say `whisper --model tiny.en`, is it using the model from HF?

**Reply by zhuzilin** [reply]
@abhinavkulkarni yes, it is.

**Reply by arshadshk** [reply]
@zhuzilin any docs available for inference using the onnx versions present on hf ?

**Comment by oldwizard1010** [comment]
can you change package name from whisper to smth like whispervino?
For now it requires some tinkering to have them both which isn't nice for testing
Also you should probably change openvino to openvino-dev in requirements, because pip can't find openvino package

**Reply by Topping1** [reply]
can you please point out what you did to have them both working? I installed this package and don't know how to use the original whisper

**Reply by rjwilmsi** [reply]
@Topping1 This is what I did on Linux to rename this (from whisper to whisperov) to allow both to be installed. (It may well be that there is a shorter route as I'm not that familiar with python):

cd ~/Downloads/
git clone https://github.com/zhuzilin/whisper-openvino
cd whisper-openvino/
In setup.py make four edits to change whisper to whisperov.
In MANIFEST.in change instances of whisper to whisperov, rename whisper directory to whisperov.

pip3 install -r requirements.txt
python3 ~/Downloads/whisper-openvino/setup.py install

**Reply by Topping1** [reply]
thank you very much!

**Comment by rjwilmsi** [comment]
Looks very promising for the smaller models to make their use more feasible on lower end CPUs. Some quick numbers from me. Ryzen 5 4500U 6-core laptop CPU (yes clearly not the best option for these workloads but it's what I have). Audio of this YouTube video https://www.youtube.com/watch?v=GFu64hnqzVo, 6m 30s:

Original whisper on CPU is 6m19s on tiny.en, 15m39s on base.en, 60m45s on small.en
The openvino version is 4m20s on tiny.en, 7m45s on base.en.

So 1.5x faster on tiny and 2x on base is very helpful indeed. Note: I've found speed of whisper to be quite dependent on the audio file used, so your results may vary.

I compared the output files (tiny v tiny and base v base) and they matched exactly.

Per above, there's an issue with openvino small.en model so I can't benchmark that yet.

**Reply by thakurudit** [reply]
This is weird but I tried the same experiment on 

1. Colab
2. Macbook Air (M1 Chip)

**Model Comparison**: Tiny.en vs Tiny.en
**Audio File**: same as you

Here are the execution times (seconds):-

Colab:-

> Original Whisper: 63.137s
> OpenVINO: 141.557

Macbook:-

> Original Whisper: 36.733s
> OpenVINO: 79.804

Not sure what is happening!

**Reply by rjwilmsi** [reply]
There has since been a performance fix to original whisper - https://github.com/openai/whisper/pull/370

My numbers for above for original whisper are now 2m9s for tiny.en, 4m29s for base.en, 11m18s for small.en. So original whisper now 3x or more faster on CPU than versus my original numbers. Aside: the speed up on GPU was much less significant. I believe the comment from @tom-huntington below is that the above performance fix isn't yet in the openVINO version. If it can be implemented would be interesting to see whether openVINO still faster than current original whisper.

**Comment by genevera** [comment]
I'm seeing the same dynamic shape error as others except with openvino 2022.3.  Any ideas?

```bash
 2d [genevera:~/src/ml/ovarm/openvino/build/wheels ⚞2] [openvino-whisper] [2.7.4] master 2s 1 ± pip list
Package            Version
------------------ ------------------
certifi            2022.9.24
charset-normalizer 2.1.1
clang              12.0.1
ffmpeg-python      0.2.0
filelock           3.8.0
future             0.18.2
huggingface-hub    0.10.1
idna               3.4
more-itertools     8.14.0
numpy              1.23.1
openvino           2022.3.0
packaging          21.3
Pillow             9.2.0
pip                22.3
pyparsing          3.0.9
PyYAML             6.0
regex              2022.9.13
requests           2.28.1
semantic-version   2.10.0
setuptools         65.5.0
setuptools-rust    1.5.2
tokenizers         0.13.1
torch              1.12.1
torchaudio         0.13.0.dev20221017
torchvision        0.15.0.dev20221017
tqdm               4.64.1
transformers       4.23.1
typing_extensions  4.4.0
urllib3            1.26.12
wheel              0.37.1
whisper            1.0

⌁ 2d [genevera:~/src/ml/ovarm/openvino/build/wheels ⚞2] [openvino-whisper] [2.7.4] master 6s ± whisper --model base.en --language en ./tpn.wav
Traceback (most recent call last):
  File "/Users/genevera/.pyenv/versions/openvino-whisper/bin/whisper", line 8, in <module>
    sys.exit(cli())
  File "/Users/genevera/.pyenv/versions/3.9.13/envs/openvino-whisper/lib/python3.9/site-packages/whisper/transcribe.py", line 283, in cli
    model = load_model(model_name)
  File "/Users/genevera/.pyenv/versions/3.9.13/envs/openvino-whisper/lib/python3.9/site-packages/whisper/__init__.py", line 104, in load_model
    model = Whisper(dims, name)
  File "/Users/genevera/.pyenv/versions/3.9.13/envs/openvino-whisper/lib/python3.9/site-packages/whisper/model.py", line 77, in __init__
    self.decoder = OpenVinoTextDecoder(model=model)
  File "/Users/genevera/.pyenv/versions/3.9.13/envs/openvino-whisper/lib/python3.9/site-packages/whisper/model.py", line 55, in __init__
    self.model = self.core.compile_model(self._model, "CPU")
  File "/Users/genevera/.pyenv/versions/3.9.13/envs/openvino-whisper/lib/python3.9/site-packages/openvino/runtime/ie_api.py", line 386, in compile_model
    super().compile_model(model, device_name, {} if config is None else config),
RuntimeError: get_shape was called on a descriptor::Tensor with dynamic shape
```

**Reply by genevera** [reply]
bumping this - help?

**Comment by adielcahana** [comment]
Hi @zhuzilin, Thanks for this great repo.

A small suggestion, in https://github.com/zhuzilin/whisper-openvino/blob/9143b8c0508bc4366583cb941d0dd970f3fc4386/whisper/model.py#L65 I would suggest casting specifically to np.int64 as the defualt int type varies accross OS. 
We tried to use it on windows and got an error (RuntimeError: [ PARAMETER_MISMATCH ] Failed to set input blob with precision: I32, if CNNNetwork input blob precision is: I64).

Also, I tried to open an issue in the repo, but couldn't find it. I guess the option is disabled.
Thanks again!

**Comment by tom-huntington** [comment]
The caching for the audio features hasn't been ported properly. 

https://github.com/openai/whisper/blob/edb69443a541955b849a3743d567910eb023597a/whisper/model.py#L84-L86

Below is where the keys and values for `xa` are being reused in the original, but this never happens in the openVINO port.

https://github.com/openai/whisper/blob/eff383b27b783e280c089475852ba83f20f64998/whisper/model.py#L81-L83

**Comment by koudahts** [comment]
Is it possible to run on a GPU?

**Comment by jingyonghou** [comment]
> Hi, nice work!
> 
> I was immediately interested in your result, because I am working on my own custom inference implementation of Whisper for CPU and I am focusing on efficiency and low-memory usage. I just ran the same transcription that you did with the [1h 30min Cramack video](https://www.youtube.com/watch?v=P6UKhR0T6cs) and I got the following results:
> 
> | Model | Time to transcribe | Result |
> | ---      | ---                           | ---       |
> | tiny.en | 116 sec | [carmack.wav-tiny.en.txt](https://github.com/openai/whisper/files/9686673/carmack.wav-tiny.en.txt) |
> | small.en | 563 sec | [carmack.wav-small.en.txt](https://github.com/openai/whisper/files/9686680/carmack.wav-small.en.txt) |
> 
> This is running on a MacBook M1 Pro - CPU only using 8 threads.
> 
> My implementation is available here:
> 
> https://github.com/ggerganov/whisper.cpp
> 
> If you are interested, you can give it a try and see how it performs using your hardware.
> 
> ---
> 
> Edit:
> I tried installing `whisper-openvino` on my MacBook by get the following error:
> ```
> $ pip install git+https://github.com/zhuzilin/whisper-openvino.git
> Collecting git+https://github.com/zhuzilin/whisper-openvino.git
>   Cloning https://github.com/zhuzilin/whisper-openvino.git to /private/var/folders/fr/8wtnvzqx0h73rmq93vmtbv380000gn/T/pip-req-build-r430m0wd
>   Running command git clone -q https://github.com/zhuzilin/whisper-openvino.git /private/var/folders/fr/8wtnvzqx0h73rmq93vmtbv380000gn/T/pip-req-build-r430m0wd
>   Resolved https://github.com/zhuzilin/whisper-openvino.git to commit 7c38500b72d8b6ed245e59be4b303a66c56f93e0
> Requirement already satisfied: numpy in ./lib/python3.10/site-packages (from whisper==1.0) (1.23.3)
> Requirement already satisfied: torch in ./lib/python3.10/site-packages (from whisper==1.0) (1.12.1)
> Requirement already satisfied: tqdm in ./lib/python3.10/site-packages (from whisper==1.0) (4.64.1)
> Requirement already satisfied: more-itertools in ./lib/python3.10/site-packages (from whisper==1.0) (8.14.0)
> Requirement already satisfied: transformers>=4.19.0 in ./lib/python3.10/site-packages (from whisper==1.0) (4.22.2)
> Requirement already satisfied: ffmpeg-python==0.2.0 in ./lib/python3.10/site-packages (from whisper==1.0) (0.2.0)
> ERROR: Could not find a version that satisfies the requirement openvino (from whisper) (from versions: none)
> ERROR: No matching distribution found for openvino
> ```
> Maybe on Arm there are some extra steps?
> 

I have try this on a centos server, and I find your implementation will hurt the      
transcription accuracy. Did you find this problem？

**Comment by Kirankumar2609** [comment]
Has anyone tried ONNX support for whisper-large model

**Comment by koudahts** [comment]
Can someone tell me how to convert a whisper model to an openvino model?

**Reply by adrianboguszewski** [reply]
Try this tutorial: https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/227-whisper-subtitles-generation/227-whisper-subtitles-generation.ipynb

**Comment by silvacarl2** [comment]
i have not yet been able to get ONNX or openvino to work.  I know for a fact this works great though: https://github.com/ggerganov/whisper.cpp

however, it does not yet have an API wrapper, we haev not figured out yet or had the time to do that.

**Comment by albert-id** [comment]
Getting error on WIN 10 x64 laptop
`[ PARAMETER_MISMATCH ] Failed to set input blob with precision: I32, if CNNNetwork input blob precision is: I64`

How can i fix it? It is unable to set fp16 argument when calling model.transcribe(...) ?

**Comment by koudahts** [comment]
Specifying the model load destination to the GPU will result in a runtime error.

RuntimeError: cldnn program build failed! [GPU] get_tensor() is called for dynamic shape

Does anyone have a solution?

Thanks.

**Comment by csukuangfj** [comment]
I suggest that you also have a look at 
https://github.com/k2-fsa/sherpa-onnx/pull/238

We are supporting whisper in [sherpa-onnx](http://github.com/k2-fsa/sherpa-onnx)

At present, you can use the code from the above PR to export whisper models to ONNX and use the exported model with onnxruntime in Python for speech recognition.

We are adding C++ support to [sherpa-onnx](http://github.com/k2-fsa/sherpa-onnx)

**Comment by SchweitzerGAO** [comment]
What are the input nodes and their names? I cannot find a "readme" to the ONNX files

**Reply by SchweitzerGAO** [reply]
Or are they identical to the input of `forward` function of the encoder and decoder [here](https://github.com/openai/whisper/blob/main/whisper/model.py)?
